File: pkg/dart2cpp/sample/cpp_generated/type_inference_test.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("=== 类型推断测试 ==="));
auto intVal = dart_int(10);
auto doubleVal = dart_double(3.14);
dart_print(dart_string("int variable + double variable: not supported"));
auto result2 = dart_double(5.0)->operator_add(doubleVal);
dart_print(dart_string("literal int + double variable = ") + (result2).toString());
auto result3 = doubleVal->operator_add(dart_double(10.0));
dart_print(dart_string("double variable + literal int = ") + (result3).toString());
auto result4 = intVal->operator_add(dart_int(20));
dart_print(dart_string("int variable + int literal = ") + (result4).toString());
auto result5 = dart_int(15)->operator_add(dart_int(25));
dart_print(dart_string("int literal + int literal = ") + (result5).toString());
auto name = dart_string("World");
auto greeting = dart_string("Hello ")->operator_add(name);
dart_print(dart_string("String concat: ") + (greeting).toString());
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: pkg/dart2cpp/sample/dart/type_inference_test.dart
void main() {
  // 测试类型推断
  print("=== 类型推断测试 ===");
  
  // 整数变量 + 浮点数变量 -> 不进行自动转换，保持原类型
  int intVal = 10;
  double doubleVal = 3.14;
  // var result1 = intVal + doubleVal;  // 这应该报错，不支持混合类型变量运算
  print("int variable + double variable: not supported");
  
  // 整数字面量 + 浮点数变量 -> 字面量转换为 double
  var result2 = 5 + doubleVal;  // 5 应该被转换为 dart_double(5.0)
  print("literal int + double variable = $result2");
  
  // 浮点数变量 + 整数字面量 -> 字面量转换为 double  
  var result3 = doubleVal + 10;  // 10 应该被转换为 dart_double(10.0)
  print("double variable + literal int = $result3");
  
  // 整数变量 + 整数字面量 -> 保持 int
  var result4 = intVal + 20;
  print("int variable + int literal = $result4");
  
  // 整数字面量 + 整数字面量 -> 保持 int
  var result5 = 15 + 25;
  print("int literal + int literal = $result5");
  
  // 字符串连接
  String name = "World";
  var greeting = "Hello " + name;
  print("String concat: $greeting");
}